File: packages/sdk-python/src/agentbox/workflows.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping

from .client import AgentBoxClient, Transport
from .discovery import AgentBoxDiscovery, discover_agentbox


JsonMapping = Mapping[str, Any]


@dataclass(frozen=True)
class BootstrapAgentBoxAgentResult:
    discovery: AgentBoxDiscovery
    client: AgentBoxClient
    agent: Mapping[str, Any] | None


def bootstrap_agentbox_agent(
    *,
    base_url: str | None = None,
    agent_card_url: str | None = None,
    client: JsonMapping,
    registration: JsonMapping | None = None,
    transport: Transport | None = None,
    timeout: float | None = None,
) -> BootstrapAgentBoxAgentResult:
    discovery = discover_agentbox(
        base_url=base_url,
        agent_card_url=agent_card_url,
        transport=transport,
        timeout=timeout,
    )
    agent_client = discovery.create_client(**dict(client))
    agent = agent_client.register_agent(registration) if registration is not None else None
    return BootstrapAgentBoxAgentResult(discovery=discovery, client=agent_client, agent=None if agent is None else agent["data"])


def create_private_box_with_resources(
    client: AgentBoxClient,
    data: JsonMapping,
    **options: Any,
) -> dict[str, Any]:
    created = client.create_box(
        {
            "name": data["name"],
            **({"ttl_seconds": data["ttl_seconds"]} if data.get("ttl_seconds") is not None else {}),
            **({"metadata": data["metadata"]} if data.get("metadata") is not None else {}),
        },
        **options,
    )
    box_id = created["data"]["box"]["box_id"]
    items = []
    events = []
    artifacts = []

    for item in data.get("items") or []:
        items.append(client.put_item({"box_id": box_id, **dict(item)}, **options)["data"])

    for event in data.get("events") or []:
        events.append(client.append_event({"box_id": box_id, **dict(event)}, **options)["data"])

    for artifact in data.get("artifacts") or []:
        artifacts.append(client.attach_artifact({"box_id": box_id, **dict(artifact)}, **options)["data"])

    manifest = None
    if data.get("read_manifest", True):
        manifest = client.get_manifest({"box_id": box_id}, **options)["data"]

    return {
        "box": created["data"]["box"],
        "manifest_uri": created["data"]["manifest_uri"],
        "items": items,
        "events": events,
        "artifacts": artifacts,
        "manifest": manifest,
    }


def create_scoped_read_grant(
    client: AgentBoxClient,
    data: JsonMapping,
    **options: Any,
) -> dict[str, Any]:
    body = dict(data)
    body.setdefault("permissions", ["read"])
    return client.create_grant(body, **options)["data"]


def client_for_grant(
    client: AgentBoxClient,
    grant: JsonMapping,
    *,
    actor: str | None = None,
    identity_token: str | None = None,
    auth_scheme: str | None = None,
) -> AgentBoxClient:
    if identity_token:
        return client.with_auth(
            actor=actor,
            identity_token=identity_token,
            grant_token=str(grant["access_token"]),
            auth_scheme=auth_scheme,
        )

    return client.with_auth(
        actor=actor or str(grant["grant"]["subject"]),
        token=str(grant["access_token"]),
        auth_scheme="agentboxGrant",
    )

File: packages/sdk-js/src/workflows.ts
import { AgentBoxClient } from "./client.js";
import { discoverAgentBox, type AgentBoxDiscovery } from "./discovery.js";
import type { AgentBoxClientRequestOptions } from "./client.js";
import type {
  AgentBoxDiscoveryOptions,
  AgentBoxDiscoveredClientOptions,
  AgentBoxManifest,
  AgentBoxPermission,
  AgentIdentity,
  AppendEventInput,
  ArtifactSummary,
  AttachArtifactInput,
  BoxEventRecord,
  BoxItemRecord,
  BoxRecord,
  CreateGrantResult,
  JsonObject,
  KnownAgentRecord,
  PutItemInput,
  RegisterAgentInput
} from "./types.js";

export interface BootstrapAgentBoxAgentInput extends AgentBoxDiscoveryOptions {
  client: AgentBoxDiscoveredClientOptions;
  registration?: RegisterAgentInput;
}

export interface BootstrapAgentBoxAgentResult {
  discovery: AgentBoxDiscovery;
  client: AgentBoxClient;
  agent: KnownAgentRecord | null;
}

export interface InitialBoxItemInput extends Omit<PutItemInput, "box_id"> {}

export interface InitialBoxEventInput extends Omit<AppendEventInput, "box_id"> {}

export interface InitialBoxArtifactInput extends Omit<AttachArtifactInput, "box_id"> {}

export interface CreatePrivateBoxWorkflowInput {
  name: string;
  ttl_seconds?: number;
  metadata?: JsonObject;
  items?: InitialBoxItemInput[];
  events?: InitialBoxEventInput[];
  artifacts?: InitialBoxArtifactInput[];
  readManifest?: boolean;
}

export interface CreatePrivateBoxWorkflowResult {
  box: BoxRecord;
  manifest_uri: string;
  items: BoxItemRecord[];
  events: BoxEventRecord[];
  artifacts: ArtifactSummary[];
  manifest: AgentBoxManifest | null;
}

export interface CreateScopedReadGrantInput {
  box_id: string;
  subject: string;
  subject_identity?: AgentIdentity;
  ttl_seconds: number;
  key_prefixes?: string[];
  event_streams?: string[];
  artifact_prefixes?: string[];
  permissions?: AgentBoxPermission[];
}

export interface GrantClientOptions {
  actor?: string;
  identityToken?: string;
  authScheme?: string;
}

export async function bootstrapAgentBoxAgent(input: BootstrapAgentBoxAgentInput): Promise<BootstrapAgentBoxAgentResult> {
  const discovery = await discoverAgentBox(input);
  const client = discovery.createClient(input.client);
  const agent = input.registration ? (await client.registerAgent(input.registration)).data : null;
  return {
    discovery,
    client,
    agent
  };
}

export async function createPrivateBoxWithResources(
  client: AgentBoxClient,
  input: CreatePrivateBoxWorkflowInput,
  options?: AgentBoxClientRequestOptions
): Promise<CreatePrivateBoxWorkflowResult> {
  const created = await client.createBox(
    {
      name: input.name,
      ttl_seconds: input.ttl_seconds,
      metadata: input.metadata
    },
    options
  );
  const boxId = created.data.box.box_id;
  const items: BoxItemRecord[] = [];
  const events: BoxEventRecord[] = [];
  const artifacts: ArtifactSummary[] = [];

  for (const item of input.items ?? []) {
    items.push(
      (
        await client.putItem(
          {
            box_id: boxId,
            ...item
          },
          options
        )
      ).data
    );
  }

  for (const event of input.events ?? []) {
    events.push(
      (
        await client.appendEvent(
          {
            box_id: boxId,
            ...event
          },
          options
        )
      ).data
    );
  }

  for (const artifact of input.artifacts ?? []) {
    artifacts.push(
      (
        await client.attachArtifact(
          {
            box_id: boxId,
            ...artifact
          },
          options
        )
      ).data
    );
  }

  const manifest = input.readManifest === false ? null : (await client.getManifest({ box_id: boxId }, options)).data;

  return {
    box: created.data.box,
    manifest_uri: created.data.manifest_uri,
    items,
    events,
    artifacts,
    manifest
  };
}

export async function createScopedReadGrant(
  client: AgentBoxClient,
  input: CreateScopedReadGrantInput,
  options?: AgentBoxClientRequestOptions
): Promise<CreateGrantResult> {
  return (
    await client.createGrant(
      {
        ...input,
        permissions: input.permissions ?? ["read"]
      },
      options
    )
  ).data;
}

export function clientForGrant(
  client: AgentBoxClient,
  grant: CreateGrantResult,
  options: GrantClientOptions = {}
): AgentBoxClient {
  if (options.identityToken) {
    return client.withAuth({
      identityToken: options.identityToken,
      grantToken: grant.access_token,
      authScheme: options.authScheme,
      actor: options.actor
    });
  }

  return client.withAuth({
    actor: options.actor ?? grant.grant.subject,
    token: grant.access_token,
    authScheme: "agentboxGrant"
  });
}
